nx/lib/nx/shape.ex

```diff
@@ -1192,6 +1192,72 @@ defmodule Nx.Shape do
     {List.to_tuple(result_shape), result_names}
   end
 
+  @doc """
+  Returns the shape and names after a `take_along_axis` operation is performed.
+
+  In practice, `axis` in `shape` gets replaced by `indices_shape`.
+
+  ## Examples
+
+      iex> Nx.Shape.take_along_axis({2, 3}, {10, 3}, 0)
+      {10, 3}
+
+      iex> Nx.Shape.take_along_axis({2, 3}, {2, 10}, 1)
+      {2, 10}
+
+      iex> Nx.Shape.take_along_axis({2, 3, 4}, {10, 3, 4}, 0)
+      {10, 3, 4}
+
+  ### Error cases
+
+      iex> Nx.Shape.take_along_axis({2, 3, 4}, {3, 10, 4}, 1)
+      ** (ArgumentError) non-indexing dimensions must match. Expected {2, *, 4}, got: {3, 10, 4}
+
+      iex> Nx.Shape.take_along_axis({2, 3}, {1, 2, 3}, 0)
+      ** (ArgumentError) shapes must have the same number of dimensions. Expected {*, 3}, got: {1, 2, 3}
+  """
+  def take_along_axis(shape, indices_shape, axis) do
```

Comment:
@josevalim I'm thinking that we should just keep the input names in the output, since we keep the same overall dimensions in this implementation. WDYT?

Reply:
Agreed!